src/rollup/types.d.ts

```diff
@@ -0,0 +1,343 @@
+}
+
+export interface ModuleJSON {
+	id: string;
+	dependencies: string[];
+	code: string;
+	originalCode: string;
+	originalSourcemap: RawSourceMap | void;
+	ast: Program;
+	sourcemapChain: RawSourceMap[];
+	resolvedIds: IdMap;
+}
+
+export type ResolveIdHook = (
+	id: string,
+	parent: string
+) => Promise<string | boolean | void> | string | boolean | void;
+
+export interface TransformContext {
+	parse: (input: string, options: any) => any;
+	warn(warning: RollupWarning, pos?: { line: number; column: number }): void;
+	error(err: RollupError, pos?: { line: number; column: number }): void;
+}
+
+export type MissingExportHook = (
+	exportName: string,
+	importingModule: string,
+	importedModule: string,
+	importerStart?: number
+) => void;
+export type IsExternalHook = (
+	id: string,
+	parentId: string,
+	isResolved: boolean
+) => Promise<boolean | void> | boolean | void;
+export type LoadHook = (
+	id: string
+) => Promise<SourceDescription | string | void> | SourceDescription | string | void;
+export type TransformHook = (
+	this: TransformContext,
+	code: string,
+	id: String
+) => Promise<SourceDescription | string | void>;
+export type TransformBundleHook = (
+	code: string,
+	options: OutputOptions
+) => Promise<SourceDescription | string>;
+export type ResolveDynamicImportHook = (
+	specifier: string | Node,
+	parentId: string
+) => Promise<string | void> | string | void;
+
+export interface Plugin {
+	name: string;
+	options?: (options: InputOptions) => void;
+	load?: LoadHook;
+	resolveId?: ResolveIdHook;
+	missingExport?: MissingExportHook;
+	transform?: TransformHook;
+	transformBundle?: TransformBundleHook;
+	ongenerate?: (options: OutputOptions, source: SourceDescription) => void;
+	onwrite?: (options: OutputOptions, source: SourceDescription) => void;
+	resolveDynamicImport?: ResolveDynamicImportHook;
+
+	banner?: () => string;
+	footer?: () => string;
+	intro?: () => string;
+	outro?: () => string;
+}
+
+export interface TreeshakingOptions {
+	propertyReadSideEffects: boolean;
+	pureExternalModules: boolean;
+}
+
+export type ExternalOption = string[] | IsExternalHook;
+export type GlobalsOption = { [name: string]: string } | ((name: string) => string);
+
+export interface InputOptions {
+	input: string | string[] | { [entryAlias: string]: string };
+	manualChunks?: { [chunkAlias: string]: string[] };
+	external?: ExternalOption;
+	plugins?: Plugin[];
+
+	onwarn?: WarningHandler;
+	cache?: {
+		modules: ModuleJSON[];
+	};
+
+	acorn?: {};
+	acornInjectPlugins?: Function[];
+	treeshake?: boolean | TreeshakingOptions;
+	context?: string;
+	moduleContext?: string | ((id: string) => string) | { [id: string]: string };
+	watch?: WatcherOptions;
+	experimentalDynamicImport?: boolean;
+	experimentalCodeSplitting?: boolean;
+	preserveSymlinks?: boolean;
+	experimentalPreserveModules?: boolean;
+
+	// undocumented?
+	pureExternalModules?: boolean;
+	preferConst?: boolean;
+	perf?: boolean;
+
+	// deprecated
+	entry?: string;
```

Comment:
just a small note, for `deprecations` it's suggest to use /** @deprecated */ on top of each property that is deprecated.

Reason being is that if you have `tslint` it can warn you that you are using a deprecated property, so do the `IDE's` or any code analyse tool such as SonarCube.